radix-engine/src/engine/kernel.rs

```diff
@@ -529,72 +428,78 @@ where
         actor: REActor,
         mut call_frame_update: CallFrameUpdate,
     ) -> Result<X::Output, RuntimeError> {
-        let new_refed_nodes = self.execute_in_mode(ExecutionMode::AuthModule, |system_api| {
-            AuthModule::on_before_frame_start(&actor, &executor, system_api).map_err(|e| match e {
-                InvokeError::Error(e) => RuntimeError::ModuleError(e.into()),
-                InvokeError::Downstream(runtime_error) => runtime_error,
-            })
+        // Filter
+        self.execute_in_mode(ExecutionMode::AuthModule, |system_api| {
+            AuthModule::on_before_frame_start(&actor, &executor, system_api)
         })?;
 
-        // TODO: Do this in a better way by allowing module to execute in next call frame
-        call_frame_update.node_refs_to_copy.extend(new_refed_nodes);
-
-        for node_id in &call_frame_update.nodes_to_move {
-            self.prepare_move_downstream(*node_id, &actor)?;
+        // New Call Frame pre-processing
+        {
+            // TODO: Abstract these away
+            self.execute_in_mode(ExecutionMode::AuthModule, |system_api| {
+                AuthModule::on_call_frame_enter(&mut call_frame_update, &actor, system_api)
+            })?;
+            self.execute_in_mode(ExecutionMode::NodeMoveModule, |system_api| {
+                NodeMoveModule::on_call_frame_enter(&mut call_frame_update, &actor, system_api)
```

Comment:
Good time to merge all these `on_xxx` events into `Module` trait?

Reply:
It's not quite that simple yet...

There are two issues:
1. `dyn Module` trait object cannot have generics on method calls
2. Modules list is being used for both callee and system api so we need to come up with the right abstraction here

But yes, in the long term will be looking to have these merged :)

Reply:
Why not generics? Module is using `R: FeeReserve` generics already.

> Modules list is being used for both callee and system api

Sorry, could you elaborate on this?

Reply:
The issue is when we have a `dyn` type object accept another generic in a method call. e.g. if we add the following to the `Module` trait:
```
pub trait Module<R: FeeReserve> {
...
fn on_execute<Y: SystemApi>(&mut self, system_api: Y);
...
}
```

You get the following error: ```note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>```

Reply:
> Modules list is being used for both callee and system api

By this I mean that the `system_api` uses the module list. And we want to pass the `system_api` as we iterate through the module list, so there's a circular dependency that needs to be cleaned up.

Reply:
Thanks for the explaination. We definitely need to figure out how to do it properly.